### PR TITLE
[angular] enhancement: stricter and more precise type inference in ng.IPromise<T> and ng.IQService

### PR DESCRIPTION
**Pull request by noomorph**
Please fill in this template.

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
> Compiles on a private 1000+ TS files project.
- [x] pass the Travis CI test
> https://travis-ci.org/noomorph/DefinitelyTyped/builds/190640611
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
> Since the contribution is quite small, I haven't added my name to definition authors list.
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
> Not sure about generics misuse, but hopefully everything is okay.
- [x] Run `tsc` without errors.
> Done
- [x] Run `npm run lint package-name` if a `tslint.json` is present.
> Not present
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: 
> This pull request is inspired by more precise Promise definitions here:
https://github.com/Microsoft/TypeScript/blob/master/lib/lib.es6.d.ts#L5293
- [x] Increase the version number in the header if appropriate.
> The pull request has nothing to do with Angular's versioning, so I skip this.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "../tslint.json" }`.
> Probably, the changes are not that substantial.

### CODE REVIEW

**Comment by noomorph** [comment]
@dt-bot,

[x] passed the Travis CI test

> https://travis-ci.org/noomorph/DefinitelyTyped/builds/190640611

**Comment by mhegazy** [comment]
@smhc @andy-ms @thorn0 @dnschnur @thetric and @aluanhaddad can you review this change?

**Comment by andy-ms** [comment]
Looks good from my end.

**Comment by aluanhaddad** [comment]
This looks fantastic. I have a minor style gripe though.

**Review by aluanhaddad — CHANGES_REQUESTED** [review]
Please add spaces around the `|` operator in the added uses of union types.

**Comment by thorn0** [comment]
I wonder how much compatible with the native promises these signatures are? Would something like this work?
```ts
declare let p: Promise<number> | ng.IPromise<number>;
declare let p2: Promise<number> | ng.IPromise<number>;
p.then(x => console.log(x.toFixed()));
declare let pc: PromiseConstructor | ng.IQService;
let x: Promise<[number, number]> | ng.IPromise<number, number> = pc.all([p, p2]);
```
I'm asking because I faced such an incompatibility between the typings for the Q library and the promises Selenium WebDriver (see https://github.com/Microsoft/TypeScript/issues/11940). Would be nice if the different types of promises where more compatible so that union types could be happilly used with them.

File: angular/index.d.ts [hunk 1]
```diff
@@ -1035,13 +1035,14 @@ declare namespace angular {
...
          */
-        reject(reason?: any): IPromise<any>;
+        reject(reason?: any): IPromise<never>;
         /**
          * Wraps an object that might be a value or a (3rd party) then-able promise into a $q promise. This is useful when you are dealing with an object that might or might not be a promise, or if the promise comes from a source that can't be trusted.
          *
          * @param value Value or a promise
          */
         resolve<T>(value: IPromise<T>|T): IPromise<T>;
+        resolve<T1, T2>(value: IPromise<T1>|T2): IPromise<T1|T2>;
```
Comment (thorn0) [inline angular/index.d.ts]: Why is the proposed signature better? In `lib.es2015.promise.d.ts`, we see this:
```ts
resolve<T>(value: T | PromiseLike<T>): Promise<T>;
```
See my previous comment about compatibility between different types of promises. Their types should be structurally compatible as they all implement the same Promises/A+ spec. If the proposed signature is more accurate, then probably `lib.es2015.promise.d.ts` should be changed too.

**Comment by thorn0** [comment]
I take it that `lib.es2015.promise.d.ts` contains the type definitions for Promises/A+ compatible promises that are well-thought-through already. Why not take them as a starting point? Does `then` really need 4 type parameters?

**Comment by aluanhaddad** [comment]
@thorn0 it would be great to improve compatibility with the promise types as defined in lib.es2015.promise.d.ts but we would need to copy them here. This will cause a maintenance issue as they are currently in need of fixing, having the wrong overload order, meaning this would need to be updated again soon. I don't have the specific issue numbers handy but there are about five different issues logged on the TypeScript repo about broken type inferences for the built-in promise declarations.

**Comment by thorn0** [comment]
I see. I hoped those type definitions had already been polished to perfection. Can we probably extend them, not copy?

File: angular/index.d.ts (hunk 1 above)
Comment (noomorph) [inline angular/index.d.ts]: 1. It allows to handle rare edge cases. I agree that it is not very useful, but still there are "legal" cases that fail to compile with vanilla definitions of Promises:

```
Promise.resolve(Math.random() > 0.5 ? 2 : Promise.reject(2))
    .then(x => console.log(x), x => console.error(x));
```

2. On other hand, I agree that better if my pull request will target `lib.es2015.promise.d.ts`. Consistency across the typings is a valuable thing.

**Comment by noomorph** [comment]
@aluanhaddad, I would gladly add spaces, but that I am afraid it will interfere with the existing code style in the definitions. I see two options to make it consistent:

1) add spaces between union types everywhere in `angular.d.ts` and in the test as well.
2) remove a few spaces I've mistakenly inserted.

Would majority agree to stick to the first one?

**Comment by thorn0** [comment]
I agree if it's done in a separate PR.

File: angular/index.d.ts (hunk 1 above)
Comment (thorn0) [inline angular/index.d.ts]: Nice catch actually. This should be brought up in the TypeScript project.

**Comment by noomorph** [comment]
@thorn0 , regarding ```.then``` and four parameters — speculatively, I would say it is an edge case with a probability 0.001% to happen in a production code using Promises. I'll do my best to think of some non-synthetic examples which might benefit from this type inference and come up with ideas tomorrow. However, it might be quite hard.

Generally speaking, the type inference does not shine in cases when you have conditional branches in `successCallback` and/or `errorCallback` that conditionally return either `Promise<T1>` (or Promise.reject()) or T2 (plain value of another type). In other words, when you mix _return_ styles.

```typescript
interface IFoo1 { type: 'foo1'; f1: number; }
interface IFoo2 { type: 'foo2'; f2: number; }

var foo1: IFoo1;
var foo2: IFoo2;

var b1: boolean;
var b2: boolean;

const n: Promise<number> = Promise.resolve(b1 && b2 ? Promise.resolve(foo1) : foo2)
    // .then(() => b1 ? foo1 : Promise.reject('b1'))
    // .catch(() => b2 ? foo2 : Promise.reject('b2'))
    // .then(() => b1 ? foo1 : Promise.reject('b1'),
    //       () => b2 ? foo2 : Promise.reject('b2'))
    .then(foo => {
        return foo.type === 'foo1' ? foo.f1 : foo.f2;
    });
```

Try to run in TypeScript playground. Uncomment any line and look at the compilation errors.

My idea is to propagate to the existing Promise definitions... let's say some kind of Promise type merging, something like opening parentheses in mathematical expressions:

```
T1 | T2 = Promise<T1 | T2> = Promise<T1> | T2 = T1 | Promise<T2>,
```

Because these type equalities are pretty much valid when it comes to resolving value via `Promise.resolve()` or to returning a heterogenous result from `successCallback` or `failureCallback`, and this trick:

```
then<TResult1, TResult2>(successCallback: (promiseValue: T) => IPromise<TResult1>|TResult2...): IPromise<TResult1 | TResult2>;
```

stimulates TypeScript to infer types and merge them regardless of the fact if they are wrapped with Promise or not:

```
var b1, b2, b3, b4, b5, b6;
Promise.resolve<string | number| boolean>(b1 ?  Promise.resolve(1) : b2 ?  Promise.resolve('1') : b3 ?  Promise.resolve(true) : b4 ? (0) : b5 ?  ('0') : (false));
```

Honestly, this last example is too hard for the compiler, but what is interesting is that if I supply it with the basic types, it turns out to be enough for it to extract the rest of types from Promises.

**Comment by noomorph** [comment]
@thorn0 , could you elaborate on this please? I'm not sure to which part of conversation it applies:

> I agree if it's done in a separate PR.

Thanks.

**Comment by noomorph** [comment]
I think I got it. @thorn0, you were referring to the spaces in the file, right?

**Comment by thorn0** [comment]
> I think I got it. @thorn0, you were referring to the spaces in the file, right?

yes

**Comment by aluanhaddad** [comment]
@noomorph @thorn0 Yeah, a separate PR is a good idea. Since you are following the file's prevailing style, I withdraw my whitespace demands 😄

**Comment by aluanhaddad** [comment]
For reference, here are some issues that arise from the current promise definitions shipped by TypeScript in `lib.es2015.promise.d.ts`
https://github.com/Microsoft/TypeScript/issues/13008
https://github.com/Microsoft/TypeScript/issues/13420
https://github.com/Microsoft/TypeScript/issues/10977

Specifically, see this comment https://github.com/Microsoft/TypeScript/issues/10977#issuecomment-270848432

It looks like these will be resolved by https://github.com/Microsoft/TypeScript/pull/13487 which introduces type parameter defaults.

**Comment by noomorph** [comment]
@aluanhaddad, what do you think if I enforce in a new separate PR tslint'ing for DefinitelyTyped/angular? It's not under `tslint` currently, maybe it's worth to standardize the folder too?

@thorn0, @aluanhaddad, so, where do we decide and where should I move to? If I got what you said right, then you suggest making a PR to TypeScript definitions instead of merging this PR to DefinitelyTyped/angular, right?

**Comment by thorn0** [comment]
I think there is no harm in merging this PR now. :+1: 
As for the TS project, let's wait for Microsoft/TypeScript#13487 to get merged first.

**Comment by noomorph** [comment]
Guys, I don't want to put any pressure, but how can I facilitate the merging? It just seems like there are no objections against the PR, but at the same time, there's no movement in any direction.

If some styling/coverage issues are hindering/deterring you from merging, let me know. I still see the pending request for changes from @aluanhaddad though he withdrew his request verbally in the comments to the PR. Is that the reason that's holding us off?
 
Also, to prevent misunderstanding regarding code style in future pull requests to `DefinitelyTyped/angular` I started plugging `angular` folder into `tslint` workflow here in a separate branch: 

https://github.com/noomorph/DefinitelyTyped/blob/8e8e8b38c5033bab96501dd3a55c8a68f0092ed7/angular/tslint.json

I suppose it will be ready in a few days. If you potentially want it to be merged before you accept my PR, let me know.

Thanks in advance and don't drop out of sight if that's possible, please. :)

**Comment by thorn0** [comment]
I'm fine with merging it.
ping @mhegazy

**Comment by noomorph** [comment]
Sorry for the delay, today I'm going to fix the merge conflicts.

**Comment by noomorph** [comment]
@mhegazy, the pull request is ready to the merge again

**Comment by bowdenk7** [comment]
@noomorph thanks for the contribution!

**Pull request merged**